### PR TITLE
ceph-dencoder: COMMON - Add missing types

### PR DESCRIPTION
**Pull request by NitzanMordhai**
Currently, ceph-dencoder lacks certain common types, preventing us from accurately checking the ceph corpus for encode-decode mismatches. This pull request aims to address this issue by adding the missing types to ceph-dencoder.

To successfully incorporate these types into ceph-dencoder, we need to introduce the necessary dump and generate_test_instances functions that was missing in some types. These functions are essential for proper encode and decode of the added types.

This PR will enhance the functionality of ceph-dencoder by including the missing types, enabling a comprehensive analysis of encode-decode consistency. With the addition of these types, we can ensure the robustness and correctness of the ceph corpus.

This update will significantly contribute to improving the overall reliability and accuracy of ceph-dencoder. It allows for a more comprehensive assessment of the encode-decode behavior, leading to enhanced data integrity and stability within the ceph ecosystem.

Fixes: https://tracker.ceph.com/issues/61788

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/auth/Auth.h [hunk 1]
```diff
@@ -231,6 +275,16 @@ struct ExpiringCryptoKey {
...
   }
+  void dump(ceph::Formatter *f) const {
+    f->dump_object("key", key);
+    f->dump_stream("expiration") << expiration;
+  }
+  static void generate_test_instances(std::list<ExpiringCryptoKey*>& ls) {
+    ls.push_back(new ExpiringCryptoKey);
+    ls.push_back(new ExpiringCryptoKey);
+    ls.back()->key.set_secret(CEPH_CRYPTO_AES, bufferptr("1234567890123456", 16),
+      utime_t(123, 456));
```
Comment (rzarzynski) [inline src/auth/Auth.h:286]: nit: how about moving all the args to the newline?

File: src/auth/cephx/CephxKeyServer.h [hunk 2]
```diff
@@ -30,6 +30,10 @@ struct KeyServerData {
   /* for each service type */
   version_t rotating_ver;
   std::map<uint32_t, RotatingSecrets> rotating_secrets;
+  KeyServerData() 
+    : version(0),
```
Comment (rzarzynski) [inline src/auth/cephx/CephxKeyServer.h:34]: nit: we can initialize at member's definition, e.g.:

```cpp
version_t version{0};
```

File: src/auth/cephx/CephxKeyServer.h [hunk 3]
```diff
@@ -70,7 +74,27 @@ struct KeyServerData {
...
+    f->open_array_section("secrets");
+    for (auto& i : secrets) {
+      f->open_object_section("secret");
+      f->dump_object("service", i.first);
+      f->dump_object("auth", i.second);
+      f->close_section();
+    }
+    f->close_section();
+    f->open_array_section("rotating_secrets");
+    for (auto& i : rotating_secrets) {
```
Comment (rzarzynski) [inline src/auth/cephx/CephxKeyServer.h:89]: nit

File: src/auth/cephx/CephxProtocol.h [hunk 4]
```diff
@@ -375,6 +471,18 @@ struct CephXServiceTicketInfo {
...
+    ticket.dump(f);
+    session_key.dump(f);
+  }
+  static void generate_test_instances(std::list<CephXServiceTicketInfo*>& ls) {
+    ls.push_back(new CephXServiceTicketInfo);
+    ls.push_back(new CephXServiceTicketInfo);
+    ls.back()->ticket.global_id = 1234;
+    ls.back()->ticket.init_timestamps(utime_t(123, 456), utime_t(123, 456));
+    ls.back()->session_key.set_secret(CEPH_CRYPTO_AES, bufferptr("1234567890123456", 16),
+      utime_t(123, 456));
```
Comment (rzarzynski) [inline src/auth/cephx/CephxProtocol.h:484]: nit

File: src/cls/queue/cls_queue_ops.h [hunk 5]
```diff
@@ -47,7 +60,18 @@ struct cls_queue_enqueue_op {
...
+
+  void dump(ceph::Formatter *f) const {
+    f->dump_unsigned("data_vec_len", bl_data_vec.size());
+  }
+
+  static void generate_test_instances(std::list<cls_queue_enqueue_op*>& o) {
+    o.push_back(new cls_queue_enqueue_op);
+    o.push_back(new cls_queue_enqueue_op);
+    o.back()->bl_data_vec.push_back(ceph::buffer::list());
+    o.back()->bl_data_vec.back().append("data", 4);
```
Comment (rzarzynski) [inline src/cls/queue/cls_queue_ops.h:73]: sv

File: src/cls/queue/cls_queue_ops.h [hunk 6]
```diff
@@ -95,6 +131,29 @@ struct cls_queue_list_ret {
     decode(entries, bl);
     DECODE_FINISH(bl);
   }
+
+  void dump(ceph::Formatter *f) const {
+    f->dump_bool("is_truncated", is_truncated);
+    f->dump_string("next_marker", next_marker);
+    f->open_array_section("entries");
+    for (auto& entry : entries) {
```
Comment (rzarzynski) [inline src/cls/queue/cls_queue_ops.h:139]: const

File: src/cls/timeindex/cls_timeindex_ops.h [hunk 7]
```diff
@@ -76,6 +96,30 @@ struct cls_timeindex_list_ret {
     decode(truncated, bl);
     DECODE_FINISH(bl);
   }
+
+  void dump(ceph::Formatter *f) const {
+    f->open_array_section("entries");
+    for (std::list<cls_timeindex_entry>::const_iterator i = entries.begin();
```
Comment (rzarzynski) [inline src/cls/timeindex/cls_timeindex_ops.h:102]: range-based `for`.

File: src/cls/timeindex/cls_timeindex_types.h [hunk 8]
```diff
@@ -8,6 +8,7 @@
 #include "include/types.h"
 
 #include "include/utime.h"
+#include "common/Formatter.h"
```
Comment (rzarzynski) [inline src/cls/timeindex/cls_timeindex_types.h:11]: nit: sorting

File: src/cls/version/cls_version_ops.h [hunk 9]
```diff
@@ -44,6 +55,26 @@ struct cls_version_inc_op {
     decode(conds, bl);
     DECODE_FINISH(bl);
   }
+
+  void dump(ceph::Formatter *f) const {
+    f->dump_object("objv", objv);
+    f->open_array_section("conds");
+    for (auto& i : conds) {
```
Comment (rzarzynski) [inline src/cls/version/cls_version_ops.h:62]: const

File: src/cls/version/cls_version_ops.h [hunk 10]
```diff
@@ -66,6 +97,26 @@ struct cls_version_check_op {
...
```
Comment (rzarzynski) [inline src/cls/version/cls_version_ops.h:104]: const

File: src/cls/version/cls_version_types.cc [hunk 11]
```diff
@@ -6,12 +6,6 @@
 #include "common/ceph_json.h"
 
 
-void obj_version::dump(ceph::Formatter *f) const
```
Comment (rzarzynski) [inline src/cls/version/cls_version_types.cc:9]: ~~I'm not seeing it's being moved; really looks like a deletion.~~

File: src/common/entity_name.cc [hunk 12]
```diff
@@ -29,6 +29,25 @@ const std::array<EntityName::str_to_entity_type_t, 6> EntityName::STR_TO_ENTITY_
   { CEPH_ENTITY_TYPE_CLIENT, "client" },
 }};
 
+void EntityName::
```
Comment (rzarzynski) [inline src/common/entity_name.cc:32]: nit: let's not split this.

File: src/common/entity_name.cc [hunk 13]
```diff
@@ -29,6 +29,25 @@ const std::array<EntityName::str_to_entity_type_t, 6> EntityName::STR_TO_ENTITY_
...
+dump(ceph::Formatter *f) const
+{
+  f->dump_int("type", type);
+  f->dump_string("id", id);
+}
+
+void EntityName::
```
Comment (rzarzynski) [inline src/common/entity_name.cc:39]: nit

File: src/messages/MClientRequest.h [hunk 14]
```diff
@@ -59,6 +59,17 @@ struct SnapPayload {
     decode(metadata, iter);
     DECODE_FINISH(iter);
   }
+  void dump(ceph::Formatter *f) const {
+    for (auto &i : metadata) {
```
Comment (rzarzynski) [inline src/messages/MClientRequest.h:63]: const

**Comment by rzarzynski** [comment]
jenkins test api

File: src/auth/Auth.h [hunk 15]
```diff
@@ -95,6 +105,21 @@ struct AuthCapsInfo {
     allow_all = (bool)a;
     decode(caps, bl);
   }
+  void dump(ceph::Formatter *f) const {
+    f->dump_bool("allow_all", allow_all);
+    for (auto p = caps.begin(); !p.end(); ++p) {
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/auth/Auth.h:109]: Done

File: src/auth/Auth.h [hunk 16]
```diff
@@ -59,6 +59,16 @@ struct EntityAuth {
       decode(pending_key, bl);
     }
   }
+  void dump(ceph::Formatter *f) const {
+    f->dump_object("key", key);
+    for (auto p = caps.begin(); p != caps.end(); ++p) {
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/auth/Auth.h]: Done

File: src/auth/Auth.h [hunk 17]
```diff
@@ -295,6 +349,23 @@ struct RotatingSecrets {
   }
 
   void dump();
+  void dump(ceph::Formatter *f) const {
+    f->open_array_section("secrets");
+    for (auto& p : secrets) {
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/auth/Auth.h:353]: Done

File: src/auth/Crypto.cc [hunk 18]
```diff
@@ -511,6 +511,23 @@ void CryptoKey::decode(bufferlist::const_iterator& bl)
...
+  f->dump_int("secret.length", secret.length());
+}
+
+void CryptoKey::generate_test_instances(std::list<CryptoKey*>& ls)
+{
+  ls.push_back(new CryptoKey);
+  ls.push_back(new CryptoKey);
+  ls.back()->type = CEPH_CRYPTO_AES;
+  ls.back()->set_secret(CEPH_CRYPTO_AES, bufferptr("1234567890123456", 16),
+      utime_t(123, 456));
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/auth/Crypto.cc:527]: Done

File: src/auth/cephx/CephxKeyServer.h (hunk 2 above)
Comment (NitzanMordhai) [reply src/auth/cephx/CephxKeyServer.h:34]: Wanted to follow the exist code, but yes, i'll change that

File: src/auth/cephx/CephxKeyServer.h (hunk 3 above)
Comment (NitzanMordhai) [reply src/auth/cephx/CephxKeyServer.h:89]: Done

File: src/auth/cephx/CephxProtocol.h [hunk 19]
```diff
@@ -113,6 +139,18 @@ struct CephXTicketBlob {
...
+  void dump(ceph::Formatter *f) const {
+    f->dump_unsigned("secret_id", secret_id);
+    f->dump_unsigned("blob_len", blob.length());
+  }
+
+  static void generate_test_instances(std::list<CephXTicketBlob*>& ls) {
+    ls.push_back(new CephXTicketBlob);
+    ls.push_back(new CephXTicketBlob);
+    ls.back()->secret_id = 123;
+    ls.back()->blob.append("this is a blob", 14);
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/auth/cephx/CephxProtocol.h:148]: Done

File: src/auth/cephx/CephxProtocol.h [hunk 20]
```diff
@@ -353,6 +438,17 @@ struct CephXServiceTicket {
...
   }
+  void dump(ceph::Formatter *f) const {
+    session_key.dump(f);
+    validity.dump(f);
+  }
+  static void generate_test_instances(std::list<CephXServiceTicket*>& ls) {
+    ls.push_back(new CephXServiceTicket);
+    ls.push_back(new CephXServiceTicket);
+    ls.back()->session_key.set_secret(CEPH_CRYPTO_AES, bufferptr("1234567890123456", 16),
+      utime_t(123, 456));
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/auth/cephx/CephxProtocol.h:445]: Done

File: src/auth/cephx/CephxProtocol.h (hunk 4 above)
Comment (NitzanMordhai) [reply src/auth/cephx/CephxProtocol.h:480]: Done

File: src/cls/2pc_queue/cls_2pc_queue_ops.h [hunk 21]
```diff
@@ -61,6 +83,24 @@ struct cls_2pc_queue_commit_op {
     DECODE_FINISH(bl);
   }
 
+  void dump(ceph::Formatter *f) const {
+    f->dump_unsigned("id", id);
+    f->open_array_section("bl_data_vec");
+    for (const auto& bl : bl_data_vec) {
+      f->dump_unsigned("bl_data_vec", bl.length());
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/cls/2pc_queue/cls_2pc_queue_ops.h:90]: no, but changed to encode_json to dump all of it

File: src/cls/2pc_queue/cls_2pc_queue_types.h [hunk 22]
```diff
@@ -68,5 +80,28 @@ struct cls_2pc_urgent_data
     }
     DECODE_FINISH(bl);
   }
+
+  void dump(ceph::Formatter *f) const {
+    f->dump_unsigned("reserved_size", reserved_size);
+    f->dump_unsigned("last_id", last_id);
+    f->open_array_section("reservations");
+    for (auto& [id, res] : reservations) {
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/cls/2pc_queue/cls_2pc_queue_types.h:88]: Done

File: src/cls/log/cls_log_ops.h [hunk 23]
```diff
@@ -81,6 +96,32 @@ struct cls_log_list_ret {
     decode(truncated, bl);
     DECODE_FINISH(bl);
   }
+
+  void dump(ceph::Formatter* f) const {
+    f->open_array_section("entries");
+    for (std::list<cls_log_entry>::const_iterator iter = entries.begin();
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/cls/log/cls_log_ops.h:102]: Done

File: src/cls/queue/cls_queue_ops.h (hunk 5 above)
Comment (NitzanMordhai) [reply src/cls/queue/cls_queue_ops.h:73]: Done

File: src/cls/queue/cls_queue_ops.h [hunk 24]
```diff
@@ -95,6 +131,29 @@ struct cls_queue_list_ret {
...
+
+  static void generate_test_instances(std::list<cls_queue_list_ret*>& o) {
+    o.push_back(new cls_queue_list_ret);
+    o.push_back(new cls_queue_list_ret);
+    o.back()->is_truncated = true;
+    o.back()->next_marker = "foo";
+    o.back()->entries.push_back(cls_queue_entry());
+    o.back()->entries.push_back(cls_queue_entry());
+    o.back()->entries.back().marker = "id";
+    o.back()->entries.back().data.append("data", 4);
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/cls/queue/cls_queue_ops.h:155]: Done

File: src/cls/queue/cls_queue_ops.h (hunk 6 above)
Comment (NitzanMordhai) [reply src/cls/queue/cls_queue_ops.h:139]: Done

File: src/cls/timeindex/cls_timeindex_ops.h (hunk 7 above)
Comment (NitzanMordhai) [reply src/cls/timeindex/cls_timeindex_ops.h:102]: Done

File: src/cls/timeindex/cls_timeindex_types.h (hunk 8 above)
Comment (NitzanMordhai) [reply src/cls/timeindex/cls_timeindex_types.h:11]: Done

File: src/cls/version/cls_version_ops.h (hunk 9 above)
Comment (NitzanMordhai) [reply src/cls/version/cls_version_ops.h:62]: Done

File: src/common/entity_name.cc (hunk 12 above)
Comment (NitzanMordhai) [reply src/common/entity_name.cc:32]: sure

File: src/messages/MClientRequest.h (hunk 14 above)
Comment (NitzanMordhai) [reply src/messages/MClientRequest.h:63]: Done

**Comment by NitzanMordhai** [comment]
jenkins retest this please

**Comment by NitzanMordhai** [comment]
jenkins test make check

**Comment by NitzanMordhai** [comment]
jenkins test make check

**Comment by NitzanMordhai** [comment]
jenkins retest this please

**Comment by athanatos** [comment]
@NitzanMordhai Looks like @rzarzynski approved this PR -- is it still relevant?

File: src/cls/version/cls_version_types.h [hunk 25]
```diff
@@ -53,7 +53,11 @@ struct obj_version {
             tag.compare(v.tag) == 0);
   }
 
-  void dump(ceph::Formatter *f) const;
+  void dump(ceph::Formatter *f) const {
```
Comment (rzarzynski) [inline src/cls/version/cls_version_types.h:56]: ACK (for the comment on `obj_version` from the last round).

**Review by rzarzynski — APPROVED** [review]

**Comment by rzarzynski** [comment]
@athanatos: Yup, we want those changes.

**Comment by cbodley** [comment]
has any of this been backported to reef? ultimately we want to enable the 18.2.0 corpus there too so we can catch encoding regressions in reef backports

**Pull request merged**